Create Godot Meta Toolkit Demo Project

devloglogan opened the pull request:
Update the demo project of the repo to showcase a few features of the platform SDK

- Achievements (Simple, Count, Bitfield)
- Downloadable content (Free DLC containing text to display)
- Display bidirectional followers that also own the app
- Display user name / profile image
- Display SDK initialization status and user entitlement status

dsnopek commented on demo/export_presets.cfg:
```diff
@@ -9,7 +9,7 @@ custom_features=""
 include_filter=""
 exclude_filter=""
-export_path=""
+export_path="../../misc/gdmetatoolkitstuff/Godot Meta Toolkit Demo.apk"
```
This a path outside the project. This should be left empty or made into a path in the project that we know the developer would have

devloglogan replied:
I've removed this

dsnopek commented on demo/main.gd:
```diff
@@ -1,46 +1,348 @@
+func initialize_platform_sdk():
+	var result: MetaPlatformSDK_Message
+
+	result = await MetaPlatformSDK.initialize_platform_async("8791910257582192").completed
```
We need some way to avoid including our app ID in the repo, and making it clear to developers that they need to provide their own app ID if they want to build and test this project themselves.

A strategy that I've used in the past is having some code like this:

```
# You need to supply your own application ID from https://developers.meta.com/ in order to test this app.
var APPLICATION_ID = ''

# ... later on in _ready()...
func _ready():
    if ResourceLoader.exists("res://local.gd"):
        var local = load('res://local.gd')
        if local and "APPLICATION_ID" in local:
            APPLICATION_ID = local.APPLICATION_ID
```

... then we can have a file called `local.gd` that's in `.gitignore` (so it doesn't get committed) which has our actual app ID.

But there's a whole bunch of way to do it.

devloglogan replied:
Good point, I've applied this to the project!

dsnopek commented on demo/main.gd:
```diff
@@ -1,46 +1,348 @@
-		hand_tracking_source.resize(OpenXRInterface.HAND_MAX)
-		for hand in OpenXRInterface.HAND_MAX:
-			hand_tracking_source[hand] = xr_interface.get_hand_tracking_source(hand)
+	OS.request_permissions()
```
Do we need to wait for the permissions to be granted before starting? If we make the requests that need this automatically, then we would. If the requests only happen due to user input, then we don't because the user would had to have gotten past the permission screen in order to do the interaction.

It's the `on_request_permissions_result` signal on the scene tree, but I'm actually not sure if that only happens if do `OS.request_permission()` for each individual permission, or if `OS.request_permissions()` will trigger it too.

devloglogan replied:
I've modified it so that the related request only takes place due to user input, so this should not be an issue anymore!

dsnopek commented on demo/addons/godotopenxrvendors/.gitignore:
```diff
@@ -0,0 +1,2 @@
+# Ignore copies of the binaries Godot makes at startup
```
We shouldn't be including all these files from the 'godot_openxr_vendors' plugin in here. We can include that in our upload to the asset library, or if the developer pulled down this repo themselves, they should install that themselves

devloglogan replied:
I've removed them and added the addons/godotopenxrvendors directory to `.gitignore`

dsnopek reviewed (commented):
Thanks!

I haven't tested this yet, but have just some initial comments from skimming the diff. I'll have more notes later after I test it :-)

dsnopek commented on demo/main.gd:
```diff
@@ -1,46 +1,355 @@
-		hand_tracking_source.resize(OpenXRInterface.HAND_MAX)
-		for hand in OpenXRInterface.HAND_MAX:
-			hand_tracking_source[hand] = xr_interface.get_hand_tracking_source(hand)
+	if ResourceLoader.exists("res://local.gd"):
+		var local = load('res://local.gd')
+		if local and "APPLICATION_ID" in local:
+			APPLICATION_ID = local.APPLICATION_ID
```
I'm not sure the reason, but right now the app just seems to crash if I don't have a `local.gd` file with the application ID. Better would be show some kind of message in the app tell developers what to do if the application ID is missing.

When I added the `local.gd` file, though, it worked fine

devloglogan replied:
I believe the crash shouldn't happen now. It now also indicates if no ID is provided!

dsnopek commented on demo/main.gd:
```diff
@@ -1,46 +1,355 @@
+	for friend in friends:
+		var friend_name = friend.display_name if friend.display_name != "" else friend.oculus_id
+		friend_names_label.text += friend_name + "\n"
+	friend_names_label.text = friend_names_label.text.reverse().erase(0, 2).reverse()
```
I think something might not be right here.  Under "Friends Who Own This App", I've got "Loga" (no "n"), which I think is probably due to this line of code removing the last couple characters of the friend name. Why is it doing this?

devloglogan replied:
Haha, I was attempting to remove the final newline of the string but failed to realize that `\n` counts as just a single character. I've now realized the way cleaner way of doing this would be `String.left(-1)`. :) However, there's no real need for this anyway. I've adjusted the label position and removed it.

dsnopek commented on demo/main.gd:
```diff
@@ -1,46 +1,355 @@
+	if friend_array.size() == 0:
+		return
+
+	var friends: Array[MetaPlatformSDK_User] = []
+	for friend in friend_array:
+		if friends.size() >= MAX_DISPLAY_FRIENDS:
+			break
+		friends.append(friend)
```
Rather than making a new array, why not just track a count and break out of the loop if it exceeds the max? Transferring to a Godot array kinda gives the impression that there is something wrong with the array from the SDK.

```
var friend_count := 0
for friend in friend_array:
    # ... update the label text ...
    friend_count += 1
    if friend_count >= MAX_DISPLAY_FRIENDS:
        break
```

devloglogan replied:
Done!

dsnopek commented on demo/main.gd:
```diff
@@ -1,46 +1,355 @@
+	dlc_label.text = "Processing..."
+	dlc_processing = true
+
+	var result: MetaPlatformSDK_Message = await MetaPlatformSDK.iap_launch_checkout_flow_async("00").completed
```
Could you use a well named variable or a comment to denote what the `"00"` is? Like:

```suggestion
	var product_sku := "00"
	var result: MetaPlatformSDK_Message = await MetaPlatformSDK.iap_launch_checkout_flow_async(product_sku).completed
```

... or, maybe use a constant like you're doing for the acheivements?

Also, when testing, I could see the panel for the IAP flow come up but not populate with any information - it just went away right away, and then I seemed to have the DLC. Is this what is supposed to happen? Is this a function of this being an app in development?

devloglogan replied:
I've put the SKU in a constant up top like the others.

The panel pop up is definitely caused by the `MetaPlatformSDK.iap_launch_checkout_flow_async()` function. I'm pretty positive the reason the panel doesn't actually do anything is because the DLC is free. I believe if the DLC is free it's actually possible to download it without entering the checkout flow at all, as the asset details could be `free` rather than `entitled`/`not_entitled`.

I'll verify this and remove the checkout flow function if this works as expected.

dsnopek replied:
I think developers will be primarily interested in how this works for paid stuff, actually. So, it probably makes to keep the code demonstrating that, even if it's a little weird in the app. We could just add a note about it somewhere?

devloglogan replied:
Ok! I've added a comment above the launch checkout flow function mentioning this.

dsnopek commented on demo/main.gd:
```diff
@@ -1,46 +1,355 @@
+
+	var image_request = HTTPRequest.new()
+	add_child(image_request)
+	image_request.request_completed.connect(self._image_request_completed)
```
This isn't actually important, but you never clean up the `HTTPRequest` node which you probably should to demonstrate best practices :-)

I usually do this in the function connected to `request_completed` - you can do bind to pass the node object along:

```suggestion
	image_request.request_completed.connect(self._image_request_completed.bind(image_request))
```

... and then `queue_free()` it there

devloglogan replied:
Added!

dsnopek reviewed (commented):
I just tested, and this seems to work great! Just a handful of additional notes

dsnopek commented:
Another thing I just thought of: we should put something in the documentation referring to this sample, so that folks know its there. My plan is to put the sample in the asset library as well, so we could mention/link that too (although, we won't have the link URL until after I publish it in the asset library).

devloglogan commented on demo/export_presets.cfg:
```diff
@@ -124,13 +124,13 @@ permissions/install_location_provider=false
 permissions/manage_app_tokens=false
 permissions/manage_documents=false
-permissions/manage_external_storage=false
+permissions/manage_external_storage=true
```
Unfortunately, this permission does seem required. If `READ_EXTERNAL_STORAGE` (which is noted as deprecated in the editor) is the only permission selected, `OS.request_permissions()` does provide a less "scary" prompt, but even after accepting it the demo project fails to open the downloaded DLC file. `MANAGE_EXTERNAL_STORAGE` makes it work with or without the `READ_EXTERNAL_STORAGE` permission, so I'll remove the `READ_EXTERNAL_STORAGE` permission for now.

devloglogan commented:
Added a link to the `demo` directory from the getting started document. Once we have this app available on the store I figure we can link to that as well.

I've also added a readme for the demo project, which could also hold a link for the store entry when available.

devloglogan commented:
I've removed the DLC example (and related permission request) and turned it purely into an IAP example where the label now just indicates if an in-app product has been purchased or not.

I've also enabled Quest 2 support.

dsnopek reviewed (approved):
Looks good to me!
